### PR TITLE
Buffer Check for SMM and MM

### PR DESCRIPTION
**Pull request by jiaxinwu**

### CODE REVIEW

File: MdeModulePkg/Universal/FaultTolerantWriteDxe/FaultTolerantWriteSmm.c
Comment (niruiyu) [inline MdeModulePkg/Universal/FaultTolerantWriteDxe/FaultTolerantWriteSmm.c:1]: The first patch only changes the name of a local function. It does not change any functionality.
The commit message should call it out clearly.

"For SMM, the SMM Handlers also can check the buffer for the Non
EDKII / OEM SMM Core + EDKII SMM driver support to make the SMM
Handlers to be more safe."
->
"For SMM, the SMM Handlers also validate the buffer to support working with
a non-EDKII SMM Core which does not validate the buffer before passing to SMM
Handlers. Such old behavior is kept."

File: MdeModulePkg/Library/VarCheckPolicyLib/VarCheckPolicyLibStandaloneMm.c [hunk 1]
```diff
@@ -31,20 +32,20 @@ VarCheckPolicyLibStandaloneConstructor (
...
 BOOLEAN
 EFIAPI
-VarCheckPolicyIsBufferOutsideValid (
+VarCheckPolicyIsCommBufferValid (
   IN EFI_PHYSICAL_ADDRESS  Buffer,
   IN UINT64                Length
   )
 {
-  return MmIsBufferOutsideMmValid (Buffer, Length);
+  return TRUE;
```
Comment (niruiyu) [inline MdeModulePkg/Library/VarCheckPolicyLib/VarCheckPolicyLibStandaloneMm.c:50]: Why is the call to MmIsBufferOutsideMmValid() removed?

File: MdeModulePkg/Library/VarCheckPolicyLib/VarCheckPolicyLibStandaloneMm.c
Comment (niruiyu) [inline MdeModulePkg/Library/VarCheckPolicyLib/VarCheckPolicyLibStandaloneMm.c:1]: For MM, the MM Handlers shall take responsibility to validate any
buffers outside MMRAM. Due to the communication buffer of
VarCheckPolicyLib must be in the MMRAM, return TRUE directly for
MM.

-->

For MM, the MM Handlers do not need to validate the
buffer if it is the CommBuffer passed from MmCore through the MmiHandler() parameter.
Return TRUE directly in this case.

File: SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c [hunk 2]
```diff
@@ -73,16 +73,28 @@ TpmNvsCommunciate (
...
+  //
+  if (!Tcg2IsPrimaryBufferValid ((UINTN)CommBuffer, TempCommBufferSize)) {
     DEBUG ((DEBUG_ERROR, "[%a] - MM Communication buffer in invalid location!\n", __func__));
     return EFI_ACCESS_DENIED;
   }
 
+  //
+  // The Secondary Buffer validation
+  //
+  if (!Tcg2IsSecondaryBufferValid (CommParams->TargetAddress, EFI_PAGES_TO_SIZE (EFI_SIZE_TO_PAGES (sizeof (TCG_NVS))))) {
```
Comment (niruiyu) [inline SecurityPkg/Tcg/Tcg2Smm/Tcg2Smm.c:89]: Secondary -> NonPrimary

File: SecurityPkg/Tcg/Tcg2Smm/Tcg2TraditionalMm.c [hunk 3]
```diff
@@ -50,14 +52,34 @@ Tcg2NotifyMmReady (
...
+  @retval FALSE This buffer is not valid.
+**/
+BOOLEAN
+Tcg2IsNonPrimaryBufferValid (
+  IN EFI_PHYSICAL_ADDRESS  Buffer,
+  IN UINT64                Length
+  )
+{
+  return TRUE;
+}
```
Comment (niruiyu) [inline SecurityPkg/Tcg/Tcg2Smm/Tcg2TraditionalMm.c:81]: should you call "  return SmmIsBufferOutsideSmmValid (Buffer, Length);" here?

Comment (jiaxinwu) [reply SecurityPkg/Tcg/Tcg2Smm/Tcg2TraditionalMm.c:81]: originally, no such logic check for traditional SMM, so, i keep the TRUE here, but i agree we can provide the correct implemention here.

**Review by niruiyu — APPROVED** [review]

**Pull request merged**